app/models/foreman_azure_rm/azure_rm.rb

```diff
@@ -104,230 +124,135 @@ def available_subnets
+      vm_name      = vm.name
+      rg_name      = vm.azure_vm.resource_group
+      os_disk      = vm.azure_vm.storage_profile.os_disk
+      data_disks   = vm.azure_vm.storage_profile.data_disks
       nic_ids      = vm.network_interface_card_ids
-      # In ARM things must be deleted in order
-      vm.destroy
-      nic_ids.each do |id|
-        nic   = azure_network_service.network_interfaces.get(id.split('/')[4],
-                                                             id.split('/')[-1])
-        ip_id = nic.public_ip_address_id
-        nic.destroy
-        if ip_id.present?
-          azure_network_service.public_ips.get(ip_id.split('/')[4],
-                                               ip_id.split('/')[-1]).destroy
+
+      sdk.delete_vm(rg_name, vm_name)
+
+      nic_ids.each do |nic_id|
+        nic = sdk.vm_nic(rg_name, nic_id.split('/')[-1])
+        if nic.present?
+          public_ip = nic.ip_configurations.first.public_ipaddress
+          sdk.delete_nic(rg_name, nic_id.split('/')[-1])
+          if public_ip.present?
+            ip_id = public_ip.id
+            sdk.delete_pip(rg_name, ip_id.split('/')[-1])
+          end
         end
       end
-      client.managed_disks.get(vm.resource_group, os_disk_name).destroy
-      data_disks.each do |disk|
-        client.managed_disks.get(vm.resource_group, disk.name).destroy
+      if os_disk.present?
+        sdk.delete_disk(rg_name, os_disk.name)
       end
-
-    rescue ActiveRecord::RecordNotFound
-      # If the VM does not exist, we don't really care.
-      true
-    end
-
-    protected
-
-    def client
-      @client ||= Fog::Compute.new(
-          :provider        => 'AzureRM',
-          :tenant_id       => tenant,
-          :client_id       => app_ident,
-          :client_secret   => secret_key,
-          :subscription_id => sub_id,
-          :environment     => 'AzureCloud'
-      )
-    end
-
-    def rg_client
-      # noinspection RubyArgCount
-      @rg_client ||= Fog::Resources::AzureRM.new(
-          tenant_id:       tenant,
-          client_id:       app_ident,
-          client_secret:   secret_key,
-          subscription_id: sub_id,
-          :environment     => 'AzureCloud'
-      )
-    end
-
-    def storage_client
-      @storage_client ||= Fog::Storage.new(
-          :provider        => 'AzureRM',
-          :tenant_id       => tenant,
-          :client_id       => app_ident,
-          :client_secret   => secret_key,
-          :subscription_id => sub_id,
-          :environment     => 'AzureCloud'
-      )
-    end
-
-    def azure_network_service
-      # noinspection RubyArgCount
-      @azure_network_service ||= Fog::Network::AzureRM.new(
-          :tenant_id       => tenant,
-          :client_id       => app_ident,
-          :client_secret   => secret_key,
-          :subscription_id => sub_id,
-          :environment     => 'AzureCloud'
-      )
+    # 2 diff observations.
```

Comment:
If you look at this latest method definition now, I could observe two things:

- First, for vm that actually exists: When I go to delete this, it deletes this from portal but throws error at foreman side, saying task rolled back, failed to save in the orchestration step and does not show anything on UI with the host entry as it is.
- Second, for vm that is already deleted, but just has an entry here in foreman as a host, gets deleted even though `vm=nil`, and in logs I can see the following exception that I am rescuing in but, still it deletes that host entry.